# Taller 1 - Modelos lineales
Catalina Polo

Ricardo Hurtado

### Ejercicio 1


In [1]:
# Importar librerias
from pathlib import Path
import pandas as pd
import tarfile
import urllib.request

import pandas as pd

try:
    from ydata_profiling import ProfileReport
except ModuleNotFoundError:
    !pip install ydata-profiling
    from ydata_profiling import ProfileReport

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.4/400.4 kB 23.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.7/296.7 kB 22.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 83.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.7/679.7 kB 41.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 105.4/105.4 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.3/43.3 kB 3.5 MB/s eta 0:00:00


In [2]:
# Cargar los datos
df = pd.read_csv("titanic.csv")

**Análisis exploratorio de los datos**

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
# Identificar y convertir variables categóricas
categorical_cols = ['Survived', 'Pclass', 'Sex', 'Embarked']
for col in categorical_cols:
    if col in df.columns:
        df[col] = df[col].astype('category')

print("Tipos de datos después de la conversión:")
df[categorical_cols].info()

Tipos de datos después de la conversión:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Survived  891 non-null    category
 1   Pclass    891 non-null    category
 2   Sex       891 non-null    category
 3   Embarked  889 non-null    category
dtypes: category(4)
memory usage: 4.1 KB


In [6]:
#Generación del resporte con ydata-profiling
profile = ProfileReport(df, title="Titanic Dataset Profile")
profile.to_notebook_iframe()

Output hidden; open in https://colab.research.google.com to view.

In [7]:
# Generar el reporte en html
profile.to_file("reporte.html")

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

**Preguntas:**

a. Cuantas filas tiene el archivo

891

b. Cuantas variables de tipo categórico tiene

Cuatro: 'Survived', 'Pclass', 'Sex', 'Embarked'


c. ¿Cuáles son las 3 palabras más comunes en la columna Name?

Son mr, miss y mrs. Conciderando que esos son títulos, los tres nombres más comunes son William, Jhon y Henry.

d. ¿Cuáles son las variables más correlacionadas con variable Survived? ¿Tiene sentido esta
correlación?

Las dos variables con mas correlación con "Survived" son: "Sex" y "Pclass". Tiene mucho sentido ya que las mujeres y las personas en clase alta tuvieron prioridad para subir a los botes salvavidas.

### Ejercicio 2

In [8]:
#Convertimos la variables categóricas a numéricas

categorical_cols = ['Survived', 'Pclass']
for col in categorical_cols:
    if col in df.columns:
        df[col] = df[col].astype('int')

print("Tipos de datos después de la conversión:")
df[categorical_cols].info()

Tipos de datos después de la conversión:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   Survived  891 non-null    int64
 1   Pclass    891 non-null    int64
dtypes: int64(2)
memory usage: 14.1 KB


In [9]:
# Convertimos la variable Sex en binaria porque observamos que tiene correlación
# con la variable "Survived"


df_temp = df.copy()

# Mapear 'male' = 0 y 'female'= 1
df_temp['Sex_numeric'] = df_temp['Sex'].map({'male': 0, 'female': 1}).astype(int)

print("DataFrame con la nueva columna 'Sex_numeric':")
display(df_temp[['Sex', 'Sex_numeric']].head())

print("\nTipos de datos después de la conversión:")
df_temp[['Sex', 'Sex_numeric']].info()

DataFrame con la nueva columna 'Sex_numeric':


Sex  Sex_numeric
0    male            0
1  female            1
2  female            1
3  female            1
4    male            0


Tipos de datos después de la conversión:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Sex          891 non-null    category
 1   Sex_numeric  891 non-null    int64   
dtypes: category(1), int64(1)
memory usage: 8.1 KB


In [10]:
# También quitamos el PassangerID porque se dio a los pasajeros de manera aleatoria y no representa una variable importante para el análisis
df_1 = df_temp.drop(columns=["PassengerId", "Sex", "Name", "Ticket", "Cabin", "Embarked"])

In [11]:
df_1.head()

Survived  Pclass   Age  SibSp  Parch     Fare  Sex_numeric
0         0       3  22.0      1      0   7.2500            0
1         1       1  38.0      1      0  71.2833            1
2         1       3  26.0      0      0   7.9250            1
3         1       1  35.0      1      0  53.1000            1
4         0       3  35.0      0      0   8.0500            0

La única categoría con valores faltantes es "Age" en un 19.8%. Como es un porcentaje alto, se van a imputar con la media por grupos definidos por el género y la clase.  

In [12]:
df_1["Age_missing"] = df_1["Age"].isna().astype(int)

df_1["age"] = df.groupby(
    ["Sex", "Pclass"]
)["Age"].transform(
    lambda x: x.fillna(x.median())
)

# Fallback global
df_1["Age"] = df_1["Age"].fillna(df_1["Age"].median())

/tmp/ipython-input-1476298480.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_1["age"] = df.groupby(


In [13]:
df_1=df_1.drop(columns=["Age_missing"])

In [14]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Survived     891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Age          891 non-null    float64
 3   SibSp        891 non-null    int64  
 4   Parch        891 non-null    int64  
 5   Fare         891 non-null    float64
 6   Sex_numeric  891 non-null    int64  
 7   age          891 non-null    float64
dtypes: float64(3), int64(5)
memory usage: 55.8 KB


In [15]:
df_1.head()

Survived  Pclass   Age  SibSp  Parch     Fare  Sex_numeric   age
0         0       3  22.0      1      0   7.2500            0  22.0
1         1       1  38.0      1      0  71.2833            1  38.0
2         1       3  26.0      0      0   7.9250            1  26.0
3         1       1  35.0      1      0  53.1000            1  35.0
4         0       3  35.0      0      0   8.0500            0  35.0

In [16]:
df_1 = df_1.drop(columns=["Age"])
df_1 = df_1.rename(columns={'age': 'Age'})

**Modelo de regresión logística**

In [17]:
# Preparación de los grupos de datos

from sklearn.model_selection import train_test_split

# Separar características y variable objetivo
X = df_1.drop('Survived', axis=1)
y = df_1['Survived'] #Objetivo

# Datos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Dimensiones de X_train:", X_train.shape)
print("Dimensiones de X_test:", X_test.shape)
print("Dimensiones de y_train:", y_train.shape)
print("Dimensiones de y_test:", y_test.shape)

Dimensiones de X_train: (712, 6)
Dimensiones de X_test: (179, 6)
Dimensiones de y_train: (712,)
Dimensiones de y_test: (179,)


In [18]:
# Entrenamiento del modelo

from sklearn.linear_model import LogisticRegression

model = LogisticRegression(random_state=42, solver='liblinear')

#Entrenmianeto
model.fit(X_train, y_train)

print(":D")

:D


In [19]:
#Evaluación del modelo

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Utilizar el conjunto de datos de prueba
y_pred = model.predict(X_test)

# Sacamos las métricas del
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

print(f"Precisión del modelo: {accuracy:.2f}")
print("\nInforme de Clasificación:\n", report)
print("\nMatriz de Confusión:\n", conf_matrix)

Precisión del modelo: 0.81

Informe de Clasificación:
               precision    recall  f1-score   support

           0       0.81      0.88      0.84       105
           1       0.80      0.72      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179


Matriz de Confusión:
 [[92 13]
 [21 53]]


Se observa un accuracy aceptable, lo que indica que, en términos generales, el modelo presenta un comportamiento adecuado. Sin embargo, si observamos en el desempeño por clase, se puede evidenciar que el modelo tiene una mayor capacidad para acertar en la predicción de la Clase 0. Esto nos muestra que el modelo es más confiable para identificar correctamente a las personas que sí podrían sobrevivir, que para detectar a a las que que no lo harían.

In [20]:
#Creamos la interfaz de Gradio

import numpy as np

def predict_survival(Pclass, Age, SibSp, Parch, Fare, Sex_numeric):
    input_data = pd.DataFrame([[Pclass, SibSp, Parch, Fare, Sex_numeric, Age]],
                              columns=X_train.columns)

    # Probabilidad de supervivencia
    prediction_proba = model.predict_proba(input_data)[:, 1]

    return float(prediction_proba[0])

print(":D")

:D


In [21]:
import gradio as gr

# Definir los componentes de entrada para Gradio
inputs = [
    gr.Slider(minimum=1, maximum=3, step=1, label="Pclass", value=3),
    gr.Slider(minimum=0.0, maximum=80.0, step=1.0, label="Age", value=25),
    gr.Slider(minimum=0, maximum=8, step=1, label="SibSp", value=0),
    gr.Slider(minimum=0, maximum=6, step=1, label="Parch", value=0),
    gr.Slider(minimum=0.0, maximum=512.0, step=0.1, label="Fare", value=30.0),
    gr.Radio([0, 1], label="Sex (0=Hombre, 1=Mujer)", value=0)
]

# Crear la interfaz Gradio
interface = gr.Interface(
    fn=predict_survival,
    inputs=inputs,
    outputs=gr.Label(label="Probabilidad de Supervivencia"),
    title="Predicción de Supervivencia en el Titanic",
    description="Ingrese los datos del pasajero para predecir la probabilidad de supervivencia.",
    live=True
)

# Lanzar la interfaz
interface.launch(inline=True, share=True)
print("Interfaz Gradio lanzada exitosamente.")

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://4d9c075db41c93a832.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


Interfaz Gradio lanzada exitosamente.


In [23]:
import joblib

joblib.dump(model, "svm_titanic_model.pkl")

['svm_titanic_model.pkl']